File: LeetCode/cpp/lc_0538.cpp
/**
 * @file     lc_0538.cpp
 * @brief    https://leetcode-cn.com/problems/convert-bst-to-greater-tree/
 * @author   YongDu
 * @date     2021-09-09
 */

/*----- 递归版 -----*/
class Solution {
public:
  int preVal{0};

  TreeNode *convertBST(TreeNode *root) {
    if (root) {
      convertBST(root->right);
      root->val += preVal;
      preVal = root->val;
      convertBST(root->left);
    }
    return root;
  }
};

/*----- 迭代版 -----*/
class Solution {
public:
  TreeNode *convertBST(TreeNode *root) {
    std::stack<TreeNode *> stk;
    TreeNode *node = root;
    int preVal = 0;

    while (node || !stk.empty()) {
      if (node) {
        stk.push(node);
        node = node->right;
      } else {
        node = stk.top();
        stk.pop();
        node->val += preVal;
        preVal = node->val;
        node = node->left;
      }
    }
    return root;
  }
};

File: LeetCode/python/lc_0538.py
"""
 * @file     lc_0538.py
 * @brief    https://leetcode-cn.com/problems/convert-bst-to-greater-tree/
 * @author   YongDu
 * @date     2021-09-09
"""
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


class Solution:
    def convertBST(self, root: TreeNode) -> TreeNode:
        def buildTree(root):
            if not root:
                return None
            buildTree(root.right)
            root.val += self.preVal
            self.preVal = root.val
            buildTree(root.left)

        self.preVal = 0
        buildTree(root)
        return root
